### PR TITLE
Send -1 for empty composition range

### PR DESCRIPTION
**Pull request by knopp**
https://github.com/flutter/flutter/issues/77600

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [X] All existing and new tests are passing.
- [ ] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Comment by stuartmorgan** [comment]
I missed that we lost this conversation in the model unification PR.

Could we get some unit tests set up for this so we don't regress it again?

**Review by cbracken — APPROVED** [review]
Thanks for the fix!

**Comment by knopp** [comment]
@cbracken, any pointers as to what the test should look like is appreciated.

**Comment by stuartmorgan** [comment]
https://github.com/flutter/engine/blob/7bc7fdb461734fff9fdf35179a00e692fb3d41fe/shell/platform/darwin/macos/framework/Source/FlutterViewControllerTest.mm has some tests that demonstrate validating calls made on a channel. You could make the text input plugin's `handleMethodCall:` visible for testing (to minimize coupling to private API), then send some synthetic message in via that and ensure that you get the expected TextInputClient.updateEditingState state object sent to the engine.

**Comment by knopp** [comment]
@stuartmorgan, @cbracken, I added a test for this. Anything else or is this good to go?

**Comment by cbracken** [comment]
Thanks @knopp. Taking a quick look. Looks like it needs a quick rebase for `shell/platform/darwin/macos/BUILD.gn` in the meantime.

File: shell/platform/darwin/macos/framework/Source/FlutterTextInputPluginTest.mm [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      [binaryMessengerMock sendOnChannel:@"flutter/textinput" message:updateCall]);
+
+  [plugin handleMethodCall:call
+                    result:^(id){
+                    }];
+
+  @try {
+    OCMVerify(  // NOLINT(google-objc-avoid-throwing-exception)
+        [binaryMessengerMock sendOnChannel:@"flutter/textinput" message:updateCall]);
+
```
Comment (cbracken) [inline shell/platform/darwin/macos/framework/Source/FlutterTextInputPluginTest.mm:78]: Nit: kill the blank line here.

**Review by cbracken — APPROVED** [review]
LGTM. Thank you for adding the test.

**Comment by cbracken** [comment]
I've kicked the build_and_test_linux presubmit. Re-run: https://cirrus-ci.com/task/6264967797145600

**Comment by chinmaygarde** [comment]
The test failure is a known flake. I've filed a report and re-run the test.

**Pull request merged**